File: 54 Spiral Matrix/spiralOrder.cpp
// Solution 2
// 2 ms
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        if( matrix.empty() ) return vector<int>();
        // number of rows       number of columns
        int m = matrix.size(), n = matrix[0].size();
        vector<int> res;
        int left = 0, right = n-1, top = 0, bot = m-1, x = 0, y = 0, dx = 1, dy = 0;
        for( int i = 0; i < m*n; i++ ){
            res.push_back( matrix[x][y] );
            x += dy, y += dx;
            if( dx == 1 && y > right ) dx = 0, dy = 1, top++, y--, x++;
            else if( dy == 1 && x > bot ) dx = -1, dy = 0, right--, x--, y--;
            else if( dx == -1 && y < left ) dx = 0, dy = -1, bot--, y++, x--;
            else if( dy == -1 && x < top ) dx = 1, dy = 0, left++, x++, y++;
        }
        return res;
    }
};


// Solution 1
// 4 ms

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        if( matrix.empty() ) return vector<int>();
        // max index of rows       max index of columns
        int m = matrix.size() - 1, n = matrix[0].size() - 1;
        vector<int> res;
        // x and y represent the upmost row_number and leftmost column_number in each spiral circle
        for( int x = 0, y = 0; x <= m && y <= n; x++, y++){
            for( int j = y; j <= n; j++ ) // works when x = m
                res.push_back( matrix[x][j]);
            for( int i = x + 1; i <= m; i++ ) // works when y == n
                res.push_back( matrix[i][n] );
            for( int j = n - 1; j >= y && x != m; j-- ) // if x == m, it has already been traversed in the first loop
                res.push_back( matrix[m][j] );
            for( int i = m - 1; i > x && y != n; i-- ) // if y == n, it has already been traversed in the second loop
                res.push_back( matrix[i][y] );
            m--;
            n--;
        }
        return res;
    }
};

File: 剑指offer/顺时针打印矩阵/spiralOrder.java
// 二维数组代替方向版
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        // 返回值
        List<Integer> res = new ArrayList<>();
        
        int m = matrix.length;
        if( m == 0 )
            return res;
        
        int n = matrix[0].length;
        if( n == 0 ) 
            return res;
        
        // 定义元素的访问状态
        boolean[][] visited = new boolean[m][n];
        
        // 定义遍历的元素索引
        int r = 0;
        int c = 0;
        // 定义方向（顺时针方向）
        int[][] direction = new int[][]{
                {0, 1},
                {1, 0},
                {0, -1},
                {-1, 0}
        };
        int di = 0;  // 当前方向索引
        
        // 遍历元素
        for( int i = 0; i < m * n ; i++ ) {
            // 添加元素
            res.add(matrix[r][c]);
            visited[r][c] = true;
            
            // 更新索引
            int nr = r + direction[di][0];
            int nc = c + direction[di][1];
            
            // 判断索引合法性
            if( nr >= 0 && nr < m && nc >= 0 && nc < n && !visited[nr][nc] ) {
                r = nr;
                c = nc;
            } else {
                // 顺时针递推到下一个方向
                di = ++di % 4;
                r = r + direction[di][0];
                c = c + direction[di][1];
            }
        }
        
        return res;
    }
    
}